### PR TITLE
Initial implementation of kvclient

### PR DESCRIPTION
**Pull request by chvck**

### CODE REVIEW

File: sdk/couchbase-core/src/authenticator.rs [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+    // fn get_client_certificate(service: ServiceType, host_port: String) ->
+    fn get_credentials(
+        &self,
+        service_type: ServiceType,
+        host_port: String,
+    ) -> CoreResult<UserPassPair>;
+}
+
+#[derive(Debug, Clone, PartialEq, Eq, Hash)]
+pub struct UserPassPair {
```
Comment (Matt-Woz) [inline sdk/couchbase-core/src/authenticator.rs:15]: Why do we need both UserPassPair & PasswordAuthenticator?

File: sdk/couchbase-core/src/kvclient.rs [hunk 2]
```diff
@@ -0,0 +1,338 @@
...
+            cli,
+            current_config: Mutex::new(config),
+            supported_features: vec![],
+            selected_bucket: Arc::new(Mutex::new(None)),
+            closed,
+        };
+
+        if should_bootstrap {
+            if let Some(b) = &bootstrap_select_bucket {
+                let mut guard = kv_cli.selected_bucket.lock().unwrap();
```
Comment (Matt-Woz) [inline sdk/couchbase-core/src/kvclient.rs:193]: NIT: More of a general thing, and we don't have to change it now, but should we prefer .expect() (w/ appropriate panic message) over .unwrap()?

**Review by Matt-Woz — APPROVED** [review]

File: sdk/couchbase-core/src/kvclient.rs (hunk 2 above)
Comment (chvck) [reply sdk/couchbase-core/src/kvclient.rs:193]: We shouldn't really be using either anywhere that we can proceed without a panic. In this case acquiring the lock only returns an error if it's been poisoned, in which case there was a panic elsewhere and we actually can't do much other than panic.

File: sdk/couchbase-core/src/authenticator.rs (hunk 1 above)
Comment (chvck) [reply sdk/couchbase-core/src/authenticator.rs:15]: UserPassPair is the return of `get_credentials` and PasswordAuthenticator is an implementation of Authenticator.

**Pull request merged**